File: src/test/java/de/fraunhofer/aisec/cpg/helpers/LocationConverterTest.java
/*
 * Copyright (c) 2021, Fraunhofer AISEC. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * 
 *      http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 *                    $$$$$$\  $$$$$$$\   $$$$$$\
 *                   $$  __$$\ $$  __$$\ $$  __$$\
 *                   $$ /  \__|$$ |  $$ |$$ /  \__|
 *                   $$ |      $$$$$$$  |$$ |$$$$\
 *                   $$ |      $$  ____/ $$ |\_$$ |
 *                   $$ |  $$\ $$ |      $$ |  $$ |
 *                   \$$$$$   |$$ |      \$$$$$   |
 *                    \______/ \__|       \______/
 *
 */
package de.fraunhofer.aisec.cpg.helpers;

import com.google.common.base.Objects;
import de.fraunhofer.aisec.cpg.sarif.PhysicalLocation;
import de.fraunhofer.aisec.cpg.sarif.Region;
import java.net.URI;
import java.util.HashMap;
import java.util.Map;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.Test;
import org.neo4j.ogm.typeconversion.CompositeAttributeConverter;

class LocationConverterTest extends BaseAttributeConverterTest<PhysicalLocation> {

  private static final String URI_STRING = "test://test:1234";
  private static final URI URI_TO_TEST = URI.create(URI_STRING);

  CompositeAttributeConverter<PhysicalLocation> getSut() {
    return new LocationConverter();
  }

  @Test
  void toEntityAttributeWithNullValueLine() {
    // arrange
    final CompositeAttributeConverter<PhysicalLocation> sut = getSut();
    // act
    final PhysicalLocation have = sut.toEntityAttribute(null);
    // assert
    Assertions.assertNull(have);
  }

  @Test
  void toEntityAttributeWithNullStartLine() {
    // arrange
    final CompositeAttributeConverter<PhysicalLocation> sut = getSut();
    final Map<String, ?> value = new HashMap<>();
    value.put(LocationConverter.START_LINE, null);
    // act
    final PhysicalLocation have = sut.toEntityAttribute(value);
    // assert
    Assertions.assertNull(have);
  }

  @Test
  void toEntityAttributeWithInteger() {
    // arrange
    final CompositeAttributeConverter<PhysicalLocation> sut = getSut();
    final Map<String, Object> value = new HashMap<>();
    final int startLineValue = 1;
    value.put(LocationConverter.START_LINE, startLineValue); // autoboxing to Integer
    final int endLineValue = 2;
    value.put(LocationConverter.END_LINE, endLineValue);
    final int startColumnValue = 3;
    value.put(LocationConverter.START_COLUMN, startColumnValue);
    final int endColumnValue = 4;
    value.put(LocationConverter.END_COLUMN, endColumnValue);
    value.put(LocationConverter.ARTIFACT, URI_STRING);
    final Region region =
        new Region(startLineValue, startColumnValue, endLineValue, endColumnValue);
    final PhysicalLocation want = new PhysicalLocation(URI_TO_TEST, region);
    // act
    final PhysicalLocation have = sut.toEntityAttribute(value);
    // assert
    Assertions.assertEquals(want, have);
  }

  @Test
  void toEntityAttributeWithLong() {
    // arrange
    final CompositeAttributeConverter<PhysicalLocation> sut = getSut();
    final Map<String, Object> value = new HashMap<>();
    final long startLineValue = 1;
    value.put(LocationConverter.START_LINE, startLineValue); // autoboxing to Long
    final long endLineValue = 2;
    value.put(LocationConverter.END_LINE, endLineValue);
    final long startColumnValue = 3;
    value.put(LocationConverter.START_COLUMN, startColumnValue);
    final long endColumnValue = 4;
    value.put(LocationConverter.END_COLUMN, endColumnValue);
    value.put(LocationConverter.ARTIFACT, URI_STRING);
    final Region region =
        new Region(
            (int) startLineValue, (int) startColumnValue,
            (int) endLineValue, (int) endColumnValue);
    final PhysicalLocation want = new PhysicalLocation(URI_TO_TEST, region);
    // act
    final PhysicalLocation have = sut.toEntityAttribute(value);
    // assert
    Assertions.assertEquals(want, have);
  }

  @Test
  void toEntityAttributeWithIntegerAndLong() {
    // arrange
    final CompositeAttributeConverter<PhysicalLocation> sut = getSut();
    final Map<String, Object> value = new HashMap<>();
    final int startLineValue = 1;
    value.put(LocationConverter.START_LINE, startLineValue); // autoboxing to Integer
    final int endLineValue = 2;
    value.put(LocationConverter.END_LINE, endLineValue);
    final long startColumnValue = 3;
    value.put(LocationConverter.START_COLUMN, startColumnValue); // autoboxing to Long
    final long endColumnValue = 4;
    value.put(LocationConverter.END_COLUMN, endColumnValue);
    value.put(LocationConverter.ARTIFACT, URI_STRING);
    final Region region =
        new Region(
            startLineValue, (int) startColumnValue,
            endLineValue, (int) endColumnValue);
    final PhysicalLocation want = new PhysicalLocation(URI_TO_TEST, region);
    // act
    final PhysicalLocation have = sut.toEntityAttribute(value);
    // assert
    Assertions.assertEquals(want, have);
  }

  @Test
  void toEntityAttributeWithStrings() {
    // arrange
    final CompositeAttributeConverter<PhysicalLocation> sut = getSut();
    final Map<String, Object> value = new HashMap<>();
    final String startLineValue = "1";
    value.put(LocationConverter.START_LINE, startLineValue);
    final String endLineValue = "2";
    value.put(LocationConverter.END_LINE, endLineValue);
    final String startColumnValue = "3";
    value.put(LocationConverter.START_COLUMN, startColumnValue);
    final String endColumnValue = "4";
    value.put(LocationConverter.END_COLUMN, endColumnValue);
    value.put(LocationConverter.ARTIFACT, URI_STRING);
    final Region region =
        new Region(
            Integer.parseInt(startLineValue),
            Integer.parseInt(startColumnValue),
            Integer.parseInt(endLineValue),
            Integer.parseInt(endColumnValue));
    final PhysicalLocation want = new PhysicalLocation(URI_TO_TEST, region);
    // act
    final PhysicalLocation have = sut.toEntityAttribute(value);
    // assert
    Assertions.assertEquals(want, have);
  }

  @Test
  void toEntityAttributeWithCustomTypes() {
    // arrange
    final CompositeAttributeConverter<PhysicalLocation> sut = getSut();
    final Map<String, Object> value = new HashMap<>();
    final Object startLineValue = new CustomNumber(1);
    value.put(LocationConverter.START_LINE, startLineValue);
    final Object endLineValue = new CustomNumber(2);
    value.put(LocationConverter.END_LINE, endLineValue);
    final Object startColumnValue = new CustomNumber(3);
    value.put(LocationConverter.START_COLUMN, startColumnValue);
    final Object endColumnValue = new CustomNumber(4);
    value.put(LocationConverter.END_COLUMN, endColumnValue);
    value.put(LocationConverter.ARTIFACT, URI_STRING);
    final Region region =
        new Region(
            Integer.parseInt(startLineValue.toString()),
            Integer.parseInt(startColumnValue.toString()),
            Integer.parseInt(endLineValue.toString()),
            Integer.parseInt(endColumnValue.toString()));
    final PhysicalLocation want = new PhysicalLocation(URI_TO_TEST, region);
    // act
    final PhysicalLocation have = sut.toEntityAttribute(value);
    // assert
    Assertions.assertEquals(want, have);
  }

  @Test
  void toEntityAttributeWithMixedTypes() {
    // arrange
    final CompositeAttributeConverter<PhysicalLocation> sut = getSut();
    final Map<String, Object> value = new HashMap<>();
    final Object startLineValue = 1;
    value.put(LocationConverter.START_LINE, startLineValue);
    final Object endLineValue = (long) 2;
    value.put(LocationConverter.END_LINE, endLineValue);
    final Object startColumnValue = "3";
    value.put(LocationConverter.START_COLUMN, startColumnValue);
    final Object endColumnValue = new CustomNumber(4);
    value.put(LocationConverter.END_COLUMN, endColumnValue);
    value.put(LocationConverter.ARTIFACT, URI_STRING);
    final Region region =
        new Region(
            Integer.parseInt(startLineValue.toString()),
            Integer.parseInt(startColumnValue.toString()),
            Integer.parseInt(endLineValue.toString()),
            Integer.parseInt(endColumnValue.toString()));
    final PhysicalLocation want = new PhysicalLocation(URI_TO_TEST, region);
    // act
    final PhysicalLocation have = sut.toEntityAttribute(value);
    // assert
    Assertions.assertEquals(want, have);
  }

  @Test
  void toEntityAttributeWithValueBiggerMaxIntBooms() {
    // arrange
    final CompositeAttributeConverter<PhysicalLocation> sut = getSut();
    final Map<String, Object> value = new HashMap<>();

    final long startLineValue = (long) Integer.MAX_VALUE + 1;
    value.put(LocationConverter.START_LINE, startLineValue);

    // act // assert
    Assertions.assertThrows(ArithmeticException.class, () -> sut.toEntityAttribute(value));
  }

  @Test
  void toEntityAttributeWithValueSmallerMinIntBooms() {
    // arrange
    final CompositeAttributeConverter<PhysicalLocation> sut = getSut();
    final Map<String, Object> value = new HashMap<>();

    final long startLineValue = (long) Integer.MIN_VALUE - 1;
    value.put(LocationConverter.START_LINE, startLineValue);

    // act // assert
    Assertions.assertThrows(ArithmeticException.class, () -> sut.toEntityAttribute(value));
  }

  @Test
  void toEntityAttributeWithAFloatBooms() {
    // arrange
    final CompositeAttributeConverter<PhysicalLocation> sut = getSut();
    final Map<String, Object> value = new HashMap<>();

    final Float startLineValue = (float) 1.;
    value.put(LocationConverter.START_LINE, startLineValue);

    // act // assert
    Assertions.assertThrows(NumberFormatException.class, () -> sut.toEntityAttribute(value));
  }

  @Test
  void toEntityAttributeWithADoubleBooms() {
    // arrange
    final CompositeAttributeConverter<PhysicalLocation> sut = getSut();
    final Map<String, Object> value = new HashMap<>();

    final Double startLineValue = 1.;
    value.put(LocationConverter.START_LINE, startLineValue);

    // act // assert
    Assertions.assertThrows(NumberFormatException.class, () -> sut.toEntityAttribute(value));
  }

  @Test
  void toEntityAttributeWithAStringBooms() {
    // arrange
    final CompositeAttributeConverter<PhysicalLocation> sut = getSut();
    final Map<String, Object> value = new HashMap<>();

    final Object startLineValue = "TEST STRING";
    value.put(LocationConverter.START_LINE, startLineValue);

    // act // assert
    Assertions.assertThrows(NumberFormatException.class, () -> sut.toEntityAttribute(value));
  }

  @Test
  void toEntityAttributeWithAObjectBooms() {
    // arrange
    final CompositeAttributeConverter<PhysicalLocation> sut = getSut();
    final Map<String, Object> value = new HashMap<>();

    final Object startLineValue = new Object();
    value.put(LocationConverter.START_LINE, startLineValue);

    // act // assert
    Assertions.assertThrows(NumberFormatException.class, () -> sut.toEntityAttribute(value));
  }

  private static class CustomNumber extends Number {

    private final int value;

    private CustomNumber(final int value) {
      this.value = value;
    }

    @Override
    public int intValue() {
      return value;
    }

    @Override
    public long longValue() {
      return intValue();
    }

    @Override
    public float floatValue() {
      return longValue();
    }

    @Override
    public double doubleValue() {
      return floatValue();
    }

    @Override
    public String toString() {
      return Integer.toString(intValue());
    }

    @Override
    public boolean equals(Object o) {
      if (this == o) return true;
      if (!(o instanceof CustomNumber)) return false;
      CustomNumber that = (CustomNumber) o;
      return value == that.value;
    }

    @Override
    public int hashCode() {
      return Objects.hashCode(value);
    }
  }
}

File: cpg-core/src/test/kotlin/de/fraunhofer/aisec/cpg/helpers/LocationConverterTest.kt
/*
 * Copyright (c) 2021, Fraunhofer AISEC. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 *                    $$$$$$\  $$$$$$$\   $$$$$$\
 *                   $$  __$$\ $$  __$$\ $$  __$$\
 *                   $$ /  \__|$$ |  $$ |$$ /  \__|
 *                   $$ |      $$$$$$$  |$$ |$$$$\
 *                   $$ |      $$  ____/ $$ |\_$$ |
 *                   $$ |  $$\ $$ |      $$ |  $$ |
 *                   \$$$$$   |$$ |      \$$$$$   |
 *                    \______/ \__|       \______/
 *
 */
package de.fraunhofer.aisec.cpg.helpers

import de.fraunhofer.aisec.cpg.persistence.CompositeAttributeConverter
import de.fraunhofer.aisec.cpg.persistence.converters.LocationConverter
import de.fraunhofer.aisec.cpg.sarif.PhysicalLocation
import de.fraunhofer.aisec.cpg.sarif.Region
import de.fraunhofer.aisec.cpg.test.*
import java.net.URI
import java.util.*
import kotlin.collections.HashMap
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertFailsWith
import kotlin.test.assertNull

internal class LocationConverterTest : BaseTest() {
    private val sut: CompositeAttributeConverter<PhysicalLocation?>
        get() {
            return LocationConverter()
        }

    @Test
    fun toEntityAttributeWithNullValueLine() {
        // arrange
        val sut: CompositeAttributeConverter<PhysicalLocation?> = sut
        // act
        val have = sut.toEntityAttribute(null)
        // assert
        assertNull(have)
    }

    @Test
    fun toEntityAttributeWithNullStartLine() {
        // arrange
        val sut: CompositeAttributeConverter<PhysicalLocation?> = sut
        val value = mutableMapOf<String, Any?>()
        value[LocationConverter.START_LINE] = null
        // act
        val have = sut.toEntityAttribute(value)
        // assert
        assertNull(have)
    }

    @Test
    fun toEntityAttributeWithInteger() {
        // arrange
        val sut: CompositeAttributeConverter<PhysicalLocation?> = sut
        val value: MutableMap<String, Any?> = HashMap()
        val startLineValue = 1
        value[LocationConverter.START_LINE] = startLineValue // autoboxing to Integer
        val endLineValue = 2
        value[LocationConverter.END_LINE] = endLineValue
        val startColumnValue = 3
        value[LocationConverter.START_COLUMN] = startColumnValue
        val endColumnValue = 4
        value[LocationConverter.END_COLUMN] = endColumnValue
        value[LocationConverter.ARTIFACT] = URI_STRING
        val region = Region(startLineValue, startColumnValue, endLineValue, endColumnValue)
        val want = PhysicalLocation(URI_TO_TEST, region)
        // act
        val have = sut.toEntityAttribute(value)
        // assert
        assertEquals(want, have)
    }

    @Test
    fun toEntityAttributeWithNullGraph() {
        val sut: CompositeAttributeConverter<PhysicalLocation?> = sut
        val have = sut.toGraphProperties(null)
        assertEquals(mapOf<String, Any?>(), have)
    }

    @Test
    fun toEntityAttributeWithIntegerGraph() {
        // arrange
        val sut: CompositeAttributeConverter<PhysicalLocation?> = sut
        val value: MutableMap<String, Any?> = HashMap()
        val startLineValue = 1
        value[LocationConverter.START_LINE] = startLineValue // autoboxing to Integer
        val endLineValue = 2
        value[LocationConverter.END_LINE] = endLineValue
        val startColumnValue = 3
        value[LocationConverter.START_COLUMN] = startColumnValue
        val endColumnValue = 4
        value[LocationConverter.END_COLUMN] = endColumnValue
        value[LocationConverter.ARTIFACT] = URI_STRING
        val region = Region(startLineValue, startColumnValue, endLineValue, endColumnValue)
        val want = PhysicalLocation(URI_TO_TEST, region)
        // act
        val have = sut.toGraphProperties(want)
        // assert
        assertEquals(value, have)
    }

    @Test
    fun toEntityAttributeWithLong() {
        // arrange
        val sut: CompositeAttributeConverter<PhysicalLocation?> = sut
        val value: MutableMap<String, Any?> = HashMap()
        val startLineValue: Long = 1
        value[LocationConverter.START_LINE] = startLineValue // autoboxing to Long
        val endLineValue: Long = 2
        value[LocationConverter.END_LINE] = endLineValue
        val startColumnValue: Long = 3
        value[LocationConverter.START_COLUMN] = startColumnValue
        val endColumnValue: Long = 4
        value[LocationConverter.END_COLUMN] = endColumnValue
        value[LocationConverter.ARTIFACT] = URI_STRING
        val region =
            Region(
                startLineValue.toInt(),
                startColumnValue.toInt(),
                endLineValue.toInt(),
                endColumnValue.toInt(),
            )
        val want = PhysicalLocation(URI_TO_TEST, region)
        // act
        val have = sut.toEntityAttribute(value)
        // assert
        assertEquals(want, have)
    }

    @Test
    fun toEntityAttributeWithIntegerAndLong() {
        // arrange
        val sut: CompositeAttributeConverter<PhysicalLocation?> = sut
        val value: MutableMap<String, Any?> = HashMap()
        val startLineValue = 1
        value[LocationConverter.START_LINE] = startLineValue // autoboxing to Integer
        val endLineValue = 2
        value[LocationConverter.END_LINE] = endLineValue
        val startColumnValue: Long = 3
        value[LocationConverter.START_COLUMN] = startColumnValue // autoboxing to Long
        val endColumnValue: Long = 4
        value[LocationConverter.END_COLUMN] = endColumnValue
        value[LocationConverter.ARTIFACT] = URI_STRING
        val region =
            Region(startLineValue, startColumnValue.toInt(), endLineValue, endColumnValue.toInt())
        val want = PhysicalLocation(URI_TO_TEST, region)
        // act
        val have = sut.toEntityAttribute(value)
        // assert
        assertEquals(want, have)
    }

    @Test
    fun toEntityAttributeWithStrings() {
        // arrange
        val sut: CompositeAttributeConverter<PhysicalLocation?> = sut
        val value: MutableMap<String, Any?> = HashMap()
        val startLineValue = "1"
        value[LocationConverter.START_LINE] = startLineValue
        val endLineValue = "2"
        value[LocationConverter.END_LINE] = endLineValue
        val startColumnValue = "3"
        value[LocationConverter.START_COLUMN] = startColumnValue
        val endColumnValue = "4"
        value[LocationConverter.END_COLUMN] = endColumnValue
        value[LocationConverter.ARTIFACT] = URI_STRING
        val region =
            Region(
                startLineValue.toInt(),
                startColumnValue.toInt(),
                endLineValue.toInt(),
                endColumnValue.toInt(),
            )
        val want = PhysicalLocation(URI_TO_TEST, region)
        // act
        val have = sut.toEntityAttribute(value)
        // assert
        assertEquals(want, have)
    }

    @Test
    fun toEntityAttributeWithCustomTypes() {
        // arrange
        val sut: CompositeAttributeConverter<PhysicalLocation?> = sut
        val value: MutableMap<String, Any?> = HashMap()
        val startLineValue: Any = CustomNumber(1)
        value[LocationConverter.START_LINE] = startLineValue
        val endLineValue: Any = CustomNumber(2)
        value[LocationConverter.END_LINE] = endLineValue
        val startColumnValue: Any = CustomNumber(3)
        value[LocationConverter.START_COLUMN] = startColumnValue
        val endColumnValue: Any = CustomNumber(4)
        value[LocationConverter.END_COLUMN] = endColumnValue
        value[LocationConverter.ARTIFACT] = URI_STRING
        val region =
            Region(
                startLineValue.toString().toInt(),
                startColumnValue.toString().toInt(),
                endLineValue.toString().toInt(),
                endColumnValue.toString().toInt(),
            )
        val want = PhysicalLocation(URI_TO_TEST, region)
        // act
        val have = sut.toEntityAttribute(value)
        // assert
        assertEquals(want, have)
    }

    @Test
    fun toEntityAttributeWithMixedTypes() {
        // arrange
        val sut: CompositeAttributeConverter<PhysicalLocation?> = sut
        val value: MutableMap<String, Any?> = HashMap()
        val startLineValue: Any = 1
        value[LocationConverter.START_LINE] = startLineValue
        val endLineValue: Any = 2L
        value[LocationConverter.END_LINE] = endLineValue
        val startColumnValue: Any = "3"
        value[LocationConverter.START_COLUMN] = startColumnValue
        val endColumnValue: Any = CustomNumber(4)
        value[LocationConverter.END_COLUMN] = endColumnValue
        value[LocationConverter.ARTIFACT] = URI_STRING
        val region =
            Region(
                startLineValue.toString().toInt(),
                startColumnValue.toString().toInt(),
                endLineValue.toString().toInt(),
                endColumnValue.toString().toInt(),
            )
        val want = PhysicalLocation(URI_TO_TEST, region)
        // act
        val have = sut.toEntityAttribute(value)
        // assert
        assertEquals(want, have)
    }

    @Test
    fun toEntityAttributeWithValueBiggerMaxIntBooms() {
        // arrange
        val sut: CompositeAttributeConverter<PhysicalLocation?> = sut
        val value: MutableMap<String, Any?> = HashMap()
        val startLineValue = Int.MAX_VALUE.toLong() + 1
        value[LocationConverter.START_LINE] = startLineValue

        // act // assert
        assertFailsWith<ArithmeticException> { sut.toEntityAttribute(value) }
    }

    @Test
    fun toEntityAttributeWithValueSmallerMinIntBooms() {
        // arrange
        val sut: CompositeAttributeConverter<PhysicalLocation?> = sut
        val value: MutableMap<String, Any?> = HashMap()
        val startLineValue = Int.MIN_VALUE.toLong() - 1
        value[LocationConverter.START_LINE] = startLineValue

        // act // assert
        assertFailsWith<ArithmeticException> { sut.toEntityAttribute(value) }
    }

    @Test
    fun toEntityAttributeWithAFloatBooms() {
        // arrange
        val sut: CompositeAttributeConverter<PhysicalLocation?> = sut
        val value: MutableMap<String, Any?> = HashMap()
        val startLineValue = 1.0.toFloat()
        value[LocationConverter.START_LINE] = startLineValue

        // act // assert
        assertFailsWith<NumberFormatException> { sut.toEntityAttribute(value) }
    }

    @Test
    fun toEntityAttributeWithADoubleBooms() {
        // arrange
        val sut: CompositeAttributeConverter<PhysicalLocation?> = sut
        val value: MutableMap<String, Any?> = HashMap()
        val startLineValue = 1.0
        value[LocationConverter.START_LINE] = startLineValue

        // act // assert
        assertFailsWith<NumberFormatException> { sut.toEntityAttribute(value) }
    }

    @Test
    fun toEntityAttributeWithAStringBooms() {
        // arrange
        val sut: CompositeAttributeConverter<PhysicalLocation?> = sut
        val value: MutableMap<String, Any?> = HashMap()
        val startLineValue: Any = "TEST STRING"
        value[LocationConverter.START_LINE] = startLineValue

        // act // assert
        assertFailsWith<NumberFormatException> { sut.toEntityAttribute(value) }
    }

    @Test
    fun toEntityAttributeWithAObjectBooms() {
        // arrange
        val sut: CompositeAttributeConverter<PhysicalLocation?> = sut
        val value: MutableMap<String, Any?> = HashMap()
        val startLineValue = Any()
        value[LocationConverter.START_LINE] = startLineValue

        // act // assert
        assertFailsWith<NumberFormatException> { sut.toEntityAttribute(value) }
    }

    private class CustomNumber(private val value: Int) : Number() {
        override fun toString(): String {
            return toInt().toString()
        }

        override fun equals(other: Any?): Boolean {
            if (this === other) return true
            if (other !is CustomNumber) return false
            return value == other.value
        }

        override fun hashCode(): Int {
            return Objects.hash(value)
        }

        override fun toByte(): Byte {
            return value.toByte()
        }

        @Deprecated(
            "Direct conversion to Char is deprecated. Use toInt().toChar() or Char constructor instead.\nIf you override toChar() function in your Number inheritor, it's recommended to gradually deprecate the overriding function and then remove it.\nSee https://youtrack.jetbrains.com/issue/KT-46465 for details about the migration",
            replaceWith = ReplaceWith("this.toInt().toChar()"),
        )
        override fun toChar(): Char {
            return value.toChar()
        }

        override fun toDouble(): Double {
            return value.toDouble()
        }

        override fun toFloat(): Float {
            return value.toFloat()
        }

        override fun toInt(): Int {
            return value
        }

        override fun toLong(): Long {
            return value.toLong()
        }

        override fun toShort(): Short {
            return value.toShort()
        }
    }

    companion object {
        private const val URI_STRING = "test://test:1234"
        private val URI_TO_TEST = URI.create(URI_STRING)
    }
}
